server/api/crud/pagination_cache.py

```diff
@@ -0,0 +1,95 @@
+
+import mlrun.common.schemas
+import mlrun.config
+import mlrun.utils.singleton
+import server.api.crud
+import server.api.utils.singletons.db
+
+
+class PaginationCache(metaclass=mlrun.utils.singleton.Singleton):
+    @staticmethod
+    def store_pagination_cache_record(
+        session: sqlalchemy.orm.Session,
+        user: str,
+        method: typing.Callable,
+        current_page: int,
+        kwargs: dict,
+    ):
+        db = server.api.utils.singletons.db.get_db()
+        return db.store_paginated_query_cache_record(
+            session, user, method.__name__, current_page, kwargs
+        )
+
+    @staticmethod
+    def get_pagination_cache_record(session: sqlalchemy.orm.Session, key: str):
+        db = server.api.utils.singletons.db.get_db()
+        return db.get_paginated_query_cache_record(session, key)
+
+    @staticmethod
+    def list_pagination_cache_records(
+        session: sqlalchemy.orm.Session,
+        key: str = None,
+        user: str = None,
+        function: str = None,
+        last_accessed_before: datetime = None,
+        order_by: typing.Optional[
+            mlrun.common.schemas.OrderType
+        ] = mlrun.common.schemas.OrderType.desc,
+    ):
+        db = server.api.utils.singletons.db.get_db()
+        return db.list_paginated_query_cache_record(
+            session, key, user, function, last_accessed_before, order_by
+        )
+
+    @staticmethod
+    def delete_pagination_cache_record(session: sqlalchemy.orm.Session, key: str):
+        db = server.api.utils.singletons.db.get_db()
+        db.delete_paginated_query_cache_record(session, key)
+
+    @staticmethod
+    def cleanup_pagination_cache(session: sqlalchemy.orm.Session):
+        db = server.api.utils.singletons.db.get_db()
+        db.list_paginated_query_cache_record(session, as_query=True).delete()
+        session.commit()
+
+    @staticmethod
+    def monitor_pagination_cache(session: sqlalchemy.orm.Session):
+        """
+        Monitor the pagination cache and remove records that are older than the cache TTL, and if the cache table
+        reached the max size, remove the oldest records.
+        """
+        cache_ttl = mlrun.config.config.httpdb.pagination_cache.ttl
+        table_max_size = mlrun.config.config.httpdb.pagination_cache.max_size
+
+        db = server.api.utils.singletons.db.get_db()
+        db.list_paginated_query_cache_record(
+            session,
+            last_accessed_before=datetime.datetime.now(datetime.timezone.utc)
+            - datetime.timedelta(seconds=cache_ttl),
+            as_query=True,
+        ).delete()
+
+        all_records_query = db.list_paginated_query_cache_record(session, as_query=True)
+        table_size = all_records_query.count()
+        if table_size > table_max_size:
+            records = all_records_query.limit(table_size - table_max_size)
+            for record in records:
+                db.delete_paginated_query_cache_record(session, record.key)
```

Comment:
that could be very slow, isnt it? why can't expose a function to delete multiple keys at once? 
e.g.: limit by 20, delete 20 queries at once?

Reply:
I tried, you get an exception from sql alchemy that you can't call delete on a query that used limit. I have no idea why it won't let us do that :(